validator/driver.go

```diff
@@ -0,0 +1,72 @@
+/*
+# Copyright 2024 NVIDIA CORPORATION
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+*/
+
+package main
+
+// driverInfo contains information about an NVIDIA driver installation.
+//
+// isHostDriver indicates whether the driver is installed directly on
+// the host at the host's root filesystem.
+//
+// hostRoot represents the host's root filesystem (typically '/').
+//
+// driverRoot and devRoot represent the absolute paths of the driver install
+// and NVIDIA device nodes on the host.
+//
+// driverRootCtrPath and devRootCtrPath represent the paths of the driver install
+// and NVIDIA device nodes in the management containers that require them, like
+// the Toolkit Container, the Device Plugin, and MIG Manager.
+type driverInfo struct {
+	isHostDriver      bool
+	hostRoot          string
+	driverRoot        string
+	driverRootCtrPath string
+	devRoot           string
+	devRootCtrPath    string
+}
+
+func getDriverInfo(isHostDriver bool, driverRoot string, hostRoot string) driverInfo {
+	if isHostDriver {
+		return driverInfo{
+			isHostDriver:      true,
+			hostRoot:          hostRoot,
+			driverRoot:        hostRoot,
+			driverRootCtrPath: "/host",
+			devRoot:           hostRoot,
+			devRootCtrPath:    "/host",
+		}
+	}
+
+	// For drivers not installed directly on the host, devRoot will
+	// either be driverRoot or hostRoot.
+	var devRoot, devRootCtrPath string
+	devRoot = root(driverRoot).getDevRoot()
```

Comment:
This is running in a container, correct. Is `driverRoot` here the container driver root or the host driver root?

Reply:
Yes this code is running in container.

Admittedly this code is a bit hard to follow. To explain using an example, let's say GPU Operator is configured with the below values:
```
hostPaths:
  rootFS: "/"
  driverInstallDir: "/run/nvidia/driver"
```
The driver-validator container will have the rootFS mounted at `/host` and the driverInstallDir mounted at `/run/nvidia/driver`.

At this point in the code, `driverRoot` will be `/run/nvidia/driver` which is both the path to the driver root in the validator container and on the host.

Reply:
For clarity, I have updated the driver validator to have two flags -- `driver-install-dir` and `driver-install-dir-ctr-path`. I have updated this code to be 
```
devRoot = root(driverInstallDirCtrPath).getDevRoot()
```
so that it is clear we are searching for the devRoot in the container.